src/components/Autocomplete/components/ComboBox/ComboBox.tsx

```diff
@@ -53,467 +42,337 @@ export interface ComboBoxProps {
+  id: idProp,
+  options,
+  selected,
+  textField,
+  preferredPosition,
+  listTitle,
+  allowMultiple,
+  actionsBefore,
+  actionsAfter,
+  contentBefore,
+  contentAfter,
+  emptyState,
+  onSelect,
+  onEndReached,
+}: ComboBoxProps) {
+  const [selectedIndex, setSelectedIndex] = useState(-1);
+  const [selectedOptions, setSelectedOptions] = useState(selected);
+  const [navigableOptions, setNavigableOptions] = useState<
+    (OptionDescriptor | ActionListItemDescriptor)[]
+  >([]);
+  const {
+    value: popoverActive,
+    setTrue: forcePopoverActiveTrue,
+    setFalse: forcePopoverActiveFalse,
+  } = useToggle(false);
+
+  const id = useUniqueId('ComboBox', idProp);
+
+  const getActionsWithIds = useCallback(
+    (
+      actions: ActionListItemDescriptor[],
+      before?: boolean,
+    ): ActionListItemDescriptor[] => {
+      if (before) {
+        return navigableOptions.slice(0, actions.length);
+      }
+      return navigableOptions.slice(-actions.length);
+    },
+    [navigableOptions],
+  );
 
-export class ComboBox extends React.PureComponent<ComboBoxProps, State> {
-  static getDerivedStateFromProps(
-    {
-      options: nextOptions,
-      selected: nextSelected,
-      actionsBefore: nextActionsBefore,
-      actionsAfter: nextActionsAfter,
-    }: ComboBoxProps,
-    {navigableOptions, selectedOptions, comboBoxId}: State,
-  ) {
-    const optionsChanged =
-      filterForOptions(navigableOptions) &&
-      nextOptions &&
-      !optionsAreEqual(navigableOptions, nextOptions);
-
-    let newNavigableOptions: (
-      | OptionDescriptor
-      | ActionListItemDescriptor
-    )[] = [];
-    if (nextActionsBefore) {
-      newNavigableOptions = newNavigableOptions.concat(nextActionsBefore);
-    }
-    if (optionsChanged || nextActionsBefore) {
-      newNavigableOptions = newNavigableOptions.concat(nextOptions);
-    }
-    if (nextActionsAfter) {
-      newNavigableOptions = newNavigableOptions.concat(nextActionsAfter);
-    }
-    newNavigableOptions = assignOptionIds(newNavigableOptions, comboBoxId);
-
-    if (optionsChanged && selectedOptions !== nextSelected) {
-      return {
-        navigableOptions: newNavigableOptions,
-        selectedOptions: nextSelected,
-      };
-    } else if (optionsChanged) {
-      return {
-        navigableOptions: newNavigableOptions,
-      };
-    } else if (selectedOptions !== nextSelected) {
-      return {selectedOptions: nextSelected};
-    }
-    return null;
-  }
-
-  state: State = {
-    comboBoxId: this.getComboBoxId(),
-    selectedOption: undefined,
-    selectedIndex: -1,
-    selectedOptions: this.props.selected,
-    navigableOptions: [],
-    popoverActive: false,
-    popoverWasActive: false,
-  };
-
-  componentDidMount() {
-    const {options, actionsBefore, actionsAfter} = this.props;
-    const comboBoxId = this.getComboBoxId();
-    let navigableOptions: (OptionDescriptor | ActionListItemDescriptor)[] = [];
+  const handlePopoverClose = useCallback(() => {
```

Comment:
`forcePopoverActiveFalse` is already a function wrapped in a useCallback, it doesn't seem very useful to repeat that again

Down below you can

```difff
- <KeypressListener keyCode={Key.Escape} handler={handlePopoverClose} />
+ <KeypressListener keyCode={Key.Escape} handler={forcePopoverActiveFalse} />
```

And delete this variable as it is no longer used